### PR TITLE
KAZOO-3367: properly handle and send updated user data to provisioner

### PR DESCRIPTION
**Pull request by fenollp**
* Fix provisioner_v5 schema validation (now data actually validates)
* Upadte provisioner_v5 API to follow the new provisioner's
* A lot of smaller stuff

### CODE REVIEW

File: applications/crossbar/src/modules/provisioner_v5.erl [hunk 1]
```diff
@@ -444,6 +436,22 @@ send_req('accounts_update', JObj, AuthToken, AccountId, _) ->
     Resp = ibrowse:send_req(UrlString, Headers, 'post', Data, HTTPOptions),
     handle_resp(Resp).
 
+payload('account', JObj, AccountId) ->
+    ResellerId = wh_services:find_reseller_id(AccountId),
+    Setters = [ fun (J) -> wh_json:set_value(<<"create_if_missing">>, true, J) end
+              , fun (J) -> wh_json:set_value(<<"reseller_id">>, ResellerId, J) end
+              , fun (J) -> wh_json:set_value(<<"data">>, JObj, J) end ],
```
Comment (k-anderson) [inline applications/crossbar/src/modules/provisioner_v5.erl]: Please add the merge flag to the account update as well.

fun (J) -> wh_json:set_value(<<"merge">>, true, J) end

File: applications/crossbar/src/modules/provisioner_v5.erl [hunk 2]
```diff
@@ -113,21 +98,74 @@ update_account(AccountId, JObj, AuthToken) ->
...
+        <<"user">> ->
+            save_user(AccountId, JObj, AuthToken);
+        <<"device">> ->
+            save_user(AccountId, JObj, AuthToken);
+        _ -> ok %% Gets rid of VMbox
+    end.
+
+-spec save_user(ne_binary(), wh_json:object(), ne_binary()) -> 'ok'.
+save_user(AccountId, JObj, AuthToken) ->
+    Realm   = wh_json:get_value(<<"realm">>, set_realm(JObj)),
```
Comment (k-anderson) [inline applications/crossbar/src/modules/provisioner_v5.erl]: Please do not add the realm to the provisioner device, this should only be present on requests to:

POST accounts/{account_id}

File: applications/crossbar/src/modules/provisioner_v5.erl [hunk 3]
```diff
@@ -113,21 +98,74 @@ update_account(AccountId, JObj, AuthToken) ->
...
+    TZ      = wh_json:get_value(<<"timezone">>, JObj),
+    OwnerId = wh_json:get_value(<<"id">>, JObj),
+    maybe_save_account(Realm, TZ, AccountId, AuthToken),
```
Comment (k-anderson) [inline applications/crossbar/src/modules/provisioner_v5.erl]: Please do not save the users timezone on the account, this should be the timezone on the account document only (if present).

File: applications/crossbar/src/modules/provisioner_v5.erl [hunk 4]
```diff
@@ -113,21 +98,74 @@ update_account(AccountId, JObj, AuthToken) ->
...
+                  {'error', _E} ->
+                      lager:debug("no endpoint for device ~s: ~p", [DeviceId,_E]);
+                  {'ok', Endpoint} ->
+                      catch save_device(Realm, TZ, Endpoint, AuthToken)
+              end
+      end, Devices).
+
+-spec get_devices_by_owner(ne_binary(), ne_binary()) -> ne_binaries().
+get_devices_by_owner(AccountDb, OwnerId) ->
+    ViewOptions = [{'startkey', [OwnerId]}
```
Comment (k-anderson) [inline applications/crossbar/src/modules/provisioner_v5.erl]: You should be able to use {'key', [OwnerId, <<"devices">>] to fetch just device document for that owner without needing to filter in the app.

File: applications/crossbar/src/modules/provisioner_v5.erl [hunk 5]
```diff
@@ -113,21 +98,74 @@ update_account(AccountId, JObj, AuthToken) ->
...
+              end
+      end, [], JObjs).
+
+-spec maybe_save_account(api_binary(), api_binary(), ne_binary(), ne_binary()) -> 'ok'.
+maybe_save_account('undefined', _, _, _) ->
+    laer:debug("cannot save account: realm is undefined");
+maybe_save_account(Realm, TZ, AccountId, AuthToken) ->
+    Setters = [ fun (J) -> wh_json:set_value(<<"realm">>, Realm, J) end
+              , fun (J) -> wh_json:set_value(<<"timezone">>, TZ, J) end ],
+    JObj = lists:foldl(fun (F,J) -> F(J) end, wh_json:new(), Setters),
```
Comment (k-anderson) [inline applications/crossbar/src/modules/provisioner_v5.erl]: Why not wh_json:from_list?

JObj = wh_json:from_list([{<<"realm">>, Realm},
                                         ,{<<"timezone">>, Tz}]),

File: applications/crossbar/src/modules/provisioner_util.erl [hunk 6]
```diff
@@ -201,9 +201,15 @@ maybe_delete_account(Context) ->
 maybe_send_contact_list(Context) ->
     maybe_send_contact_list(Context, cb_context:resp_status(Context)).
 maybe_send_contact_list(Context, 'success') ->
-    _ = case get_provisioning_type() of
+    _ = case cb_context:is_context(Context)
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/provisioner_util.erl]: Why the need for this check? This function seems to be only called from cb_accounts with a Context record.

File: applications/crossbar/src/modules/provisioner_v5.erl [hunk 7]
```diff
@@ -452,12 +434,17 @@ send_req('accounts_update', JObj, AuthToken, AccountId, _) ->
...
-    lager:debug("provisioning success ~p", [Resp]);
+    lager:debug("provisioning success ~s", [decode(Resp)]);
 handle_resp({'ok', Code, _, Resp}) ->
-    lager:warning("provisioning error ~p. ~p", [Code, Resp]);
+    lager:warning("provisioning error ~p. ~s", [Code, decode(Resp)]);
 handle_resp(_Error) ->
     lager:error("provisioning fatal error ~p", [_Error]).
 
+-spec decode(string()) -> ne_binary().
+decode(JSON) ->
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/provisioner_v5.erl]: Is this necessary? Why not just log the Resp in `handle_resp/1` ?

File: applications/crossbar/src/modules/provisioner_v5.erl [hunk 8]
```diff
@@ -444,6 +411,21 @@ send_req('accounts_update', JObj, AuthToken, AccountId, _) ->
     Resp = ibrowse:send_req(UrlString, Headers, 'post', Data, HTTPOptions),
     handle_resp(Resp).
 
+payload('account', JObj, AccountId) ->
+    ResellerId = wh_services:find_reseller_id(AccountId),
+    wh_json:from_list([ {<<"create_if_missing">>, true}
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/provisioner_v5.erl]: please ticky all atoms

File: applications/crossbar/src/modules/provisioner_v5.erl [hunk 9]
```diff
@@ -444,6 +411,21 @@ send_req('accounts_update', JObj, AuthToken, AccountId, _) ->
...
+                      , {<<"reseller_id">>, ResellerId}
+                      , {<<"merge">>, true}
+                      , {<<"data">>, JObj} ]).
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/provisioner_v5.erl]: List termintors should be dropped down a line

File: applications/crossbar/src/modules/provisioner_v5.erl [hunk 10]
```diff
@@ -444,6 +411,21 @@ send_req('accounts_update', JObj, AuthToken, AccountId, _) ->
     Resp = ibrowse:send_req(UrlString, Headers, 'post', Data, HTTPOptions),
     handle_resp(Resp).
 
+payload('account', JObj, AccountId) ->
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/provisioner_v5.erl]: Need a spec

File: applications/crossbar/src/modules/provisioner_v5.erl (hunk 7 above)
Comment (fenollp) [inline applications/crossbar/src/modules/provisioner_v5.erl]: Not necessary. Just prettier logs

**Comment by lazedo** [comment]
@fenollp @k-anderson 
provisioner_util:is_mac_address_in_use was removed. as a result we can't create new devices

**Pull request merged**